File: src/test/scala/com/yammer/metrics/tests/MeterTest.scala
package com.yammer.metrics.tests


import org.scalatest.matchers.MustMatchers
import java.util.concurrent.TimeUnit
import org.scalatest.{BeforeAndAfterEach, Spec}
import com.yammer.time.Clock
import com.yammer.metrics.Meter

class MeterTest extends Spec with MustMatchers with BeforeAndAfterEach {
  override protected def afterEach() {
    Clock.unfreezeTime()
  }

  describe("a meter with two events in a second") {
    Clock.freezeTime(100)
    val meter = new Meter
    meter.mark(2)
    Clock.freezeTime(100 + TimeUnit.NANOSECONDS.convert(1, TimeUnit.SECONDS))

    it("has a rate of two events per second") {
      meter.rate(TimeUnit.SECONDS) must be (2.0 plusOrMinus 0.0000001)
    }

    it("has two events") {
      meter.count must be(2)
    }
  }

  describe("a meter with no events") {
    val meter = new Meter

    it("has a rate of zero") {
      meter.rate(TimeUnit.DAYS) must be (0.0 plusOrMinus 0.0000001)
    }
  }

  describe("a meter with four marked events and three unmarked events") {
    val meter = new Meter
    meter.mark(3)
    meter.unmark()
    meter.unmark(2)
    meter.mark()
    
    it("has one event") {
      meter.count must be(1)
    }
  }
}

File: metrics-core/src/test/java/com/yammer/metrics/core/tests/MeterTest.java
package com.yammer.metrics.core.tests;

import com.yammer.metrics.core.Meter;
import com.yammer.metrics.core.MetricsRegistry;
import org.junit.Test;

import java.util.concurrent.TimeUnit;

import static org.hamcrest.Matchers.closeTo;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;

public class MeterTest {
    private final MetricsRegistry registry = new MetricsRegistry();
    private final Meter meter = registry.newMeter(MeterTest.class, "things", "thing", TimeUnit.SECONDS);

    @Test
    public void aBlankMeter() throws Exception {
        assertThat("the meter has a count of zero",
                   meter.getCount(),
                   is(0L));

        assertThat("the meter has a mean rate of zero",
                   meter.getMeanRate(),
                   is(closeTo(0.0, 0.001)));
    }

    @Test
    public void aMeterWithThreeEvents() throws Exception {
        meter.mark(3);

        assertThat("the meter has a count of three",
                   meter.getCount(),
                   is(3L));
    }
}
